src/ucp/rma/amo_send.c

```diff
@@ -118,52 +133,78 @@ ucs_status_ptr_t ucp_atomic_fetch_nb(ucp_ep_h ep, ucp_atomic_fetch_op_t opcode,
                                      uint64_t value, void *result, size_t op_size,
                                      uint64_t remote_addr, ucp_rkey_h rkey,
                                      ucp_send_callback_t cb)
+{
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                        UCP_OP_ATTR_FIELD_DATATYPE,
+        .datatype     = ucp_dt_make_contig(op_size),
+        .cb.send      = (ucp_send_nbx_callback_t)cb
+    };
+
+    return ucp_atomic_fetch_nbx(ep, opcode, &value, result, 1,
+                                remote_addr, rkey, &param);
+}
+
+ucs_status_ptr_t
+ucp_atomic_fetch_nbx(ucp_ep_h ep, ucp_atomic_fetch_op_t opcode,
+                     const void *buffer, void *result, size_t count,
+                     uint64_t remote_addr, ucp_rkey_h rkey,
+                     const ucp_request_param_t *param)
 {
     ucs_status_ptr_t status_p;
     ucs_status_t status;
     ucp_request_t *req;
+    uint64_t value;
+    size_t op_size;
+
+    if (ucs_likely(param->op_attr_mask & UCP_OP_ATTR_FIELD_DATATYPE)) {
+        if (param->datatype == ucp_dt_make_contig(8)) {
+            value   = *(uint64_t*)buffer;
+            op_size = sizeof(uint64_t);
+        } else if (param->datatype == ucp_dt_make_contig(4)) {
+            value   = *(uint32_t*)buffer;
+            op_size = sizeof(uint32_t);
+        } else {
+            ucs_error("invalid atomic operation datatype: %zu", param->datatype);
+            return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+        }
+    } else {
+        ucs_error("missing atomic operation datatype");
```

Comment:
can do reverse: `if (unlikely no datatype) return;` - this way no need for else

Reply:
good idea
done
thank you